### DISCUSSION TITLE
Modeling substate as an enum with associated values

### DISCUSSION THREAD

**Discussion by rzulkoski** [pointfreeco/swift-composable-architecture#1622] [Q&A] [answered]
I have a Todo list screen that is split into 3 reducers:

1. TodoList reducer (screen-level)
2. TodoListContent reducer
3. TodoListRow reducer

My TodoList.State has a `loadingState` property on it where `LoadingState` is defined as:
```Swift
/// Represents the different states of a piece of loadable substate.
public enum LoadingState<LoadedState, LoadError>: Equatable where LoadedState: Equatable, LoadError: Equatable, LoadError: Error {
  /// Substate has not yet been loaded.
  case notLoaded

  /// Substate is in the process of being loaded.
  case loading

  /// Substate has successfully loaded.
  case loaded(LoadedState)

  /// Substate has failed to load.
  case failed(LoadError)
}
```

In this instance, the `LoadedState` is `TodoListContent.State`.

I'm in the process of migrating these reducers to the ReducerProtocol and have encountered a large degradation in performance. After using some of the debugging tools available on `WithViewStore`, I have determined that any time the associated value (`LoadedState`) changes the parent view is recreating a new content view. This makes sense because the previous value of the `loadingState` property is no longer equal to the new value.

I was able to work around this issue by manually adding the equality conformance and ignoring the associated values like so:
```Swift
public enum LoadingState... {
  ... // *snip

  public var discriminator: Discriminator {
    switch self {
    case .notLoaded: return .notLoaded
    case .loading: return .loading
    case .loaded: return .loaded
    case .failed: return .failed
    }
  }

  public enum Discriminator: String, CaseIterable {
    case notLoaded, loading, loaded, failed
  }

  public static func == (lhs: Self, rhs: Self) -> Bool {
    lhs.discriminator == rhs.discriminator
  }
}
```

So here's my question: Is this a reasonable workaround for the issue I was encountering or is there a different tool I should be reaching for here?

**Comment by stephencelis** [comment, marked as answer]
@rzulkoski I think we need a little more context and code to debug the issue, but glad you found a workaround!

> I'm in the process of migrating these reducers to the ReducerProtocol and have encountered a large degradation in performance

I'm not quite sure I understand how migrating to the `ReducerProtocol` would introduce performance degradation at the view layer. Did you make changes to the view as well?

TCA provides `SwitchStore` precisely to work around the problem of enum state performance in SwiftUI by limiting updates based on the enum tag:

https://github.com/pointfreeco/swift-composable-architecture/blob/3bfbc7f9b89dccb6e7bf845dff80cb3cf7d23ba9/Sources/ComposableArchitecture/SwiftUI/SwitchStore.swift#L176

Are you using `SwitchStore`?

**Reply by rzulkoski** [reply]
> I'm not quite sure I understand how migrating to the `ReducerProtocol` would introduce performance degradation at the view layer. Did you make changes to the view as well?

I'm kind of baffled by it as well. I'm thinking I was probably doing something wrong before and just wasn't punished for it until now. Our team has been using the recommended strategy of starting at the leaf node, `TodoListRowView` in this case, and converting it while using `AnyReducer` in the parent reducer to wrap the child reducer that is being converted. We didn't notice any change when converting the row, but did have issues after migrating `TodoListContentView`.

> Are you using `SwitchStore`?

Yes. We have a helper view called `LoadableView` that uses a `SwitchStore` behind the scenes.

I'm wondering if it has to do with my outer `WithViewStore` in the code below?

```Swift
public struct TodoListView: View {
  public let store: StoreOf<TodoList>

  public init(store: StoreOf<TodoList>) {
    self.store = store
  }

  public var body: some View {
    WithViewStore(self.store) { viewStore in
      LoadableView(
        self.store,
        state: \.loadingState,
        action: TodoList.Action.contentAction,
        retryButtonTappedAction: .reload,
        selectJobButtonTappedAction: .selectJobButtonTapped
      ) { store in
        TodoListContentView(store: store)
      }
      .toolbar {
        ToolbarItem(placement: .principal) {
          BTNavigationTitleView(title: localize("toDos"), subtitle: .jobPicker)
        }

        ToolbarItem(placement: .navigationBarTrailing) {
          if viewStore.canAddTodos, viewStore.loadingState.isLoaded {
            Button(action: { viewStore.send(.addButtonTapped) }) {
              Image(systemName: "plus")
                .foregroundColor(.btOnBrandPrimary)
            }
          }
        }
      }
      .sheet(isPresented: viewStore.binding(\.$isShowingSelectAJob)) {
        SelectAJobWrapperView(onSelectJobId: { viewStore.send(.didSelectAJobForAddingToDo($0)) })
          .ignoresSafeArea()
          .interactiveDismissDisabled()
      }
      .fullScreenCover(isPresented: viewStore.binding(\.$isShowingAddToDo)) {
        TodoDetailsWrapperView(jobId: viewStore.jobIdForAddingTodo!, todoId: nil)
          .ignoresSafeArea()
      }
      .fullScreenCover(isPresented: viewStore.binding(\.$isShowingJobPicker)) {
        JobPickerWrapperView(
          allowsGeneralJob: false,
          allowsMultipleJobs: true
        )
        .ignoresSafeArea()
      }
      .alert(
        self.store.scope(state: \.errorAlert),
        dismiss: .errorAlertDismissed
      )
      .onAppear {
        viewStore.send(.onAppear)
      }
      .offlineBanner(when: viewStore.isOffline)
      .btScreenName(.toDoList)
    }.debug("**** WithViewStore.screen")
  }
}

```

**Reply by stephencelis** [reply]
Yeah, that outer `WithViewStore` is using the deprecated initializer that observes all of state automatically. See our "View Store Performance" article here:

https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/performance/#View-stores

You'll want to chip things away to just the bare essentials using a view state struct.

**Reply by rzulkoski** [reply]
What's the best way to have access to the view store from modifiers on a view without wrapping the view itself? Looking at the view code above you'll notice that I make no references to `viewStore` within `LoadableView` or its contents.

**Reply by rzulkoski** [reply]
Upon further investigation, I've determined a few things:

1. Adding the manual conformance to `Equatable` did NOT actually fix my problem.
2. Creating a `TodoListView.ViewState` did address the issue where the `TodoListContentView` was getting recreated way more than was necessary.
    - Although it was a bit awkward to include bindings (Thanks @tgrapperon for [this](https://github.com/pointfreeco/swift-composable-architecture/discussions/769#discussioncomment-1560909) helper!)
3. The real issue that was causing the strange behavior that I was seeing is that my list code is organized like this:

```Swift
public var body: some View {
  WithViewStore(self.store) { viewStore in
    List {
      Section {
        ForEachStore(self.store.scope(state: \.todos, action: TodoListContent.Action.todoRow(id:action:)) { rowStore
          TodoListRowView(store: rowStore)
        }
      } footer: {
        // Some view for representing loading more, failure to load more
      }
    }
    .listStyle(.grouped)
  }
}
```

Any time the footer needed to be refreshed, it would recreate the entire section. Is there any way around this when using `Section`? Or should I just break this up into separate views? I did like this strategy for consistently looking good across different list styles.

**Reply by fabstu** [reply]
@tgrapperon opened a discussion about such Section being refreshed in https://github.com/pointfreeco/swift-composable-architecture/discussions/1595, even when it does not use the outer viewStore. `StoreIsolated` might be useful, but haven't tried it yet. It also has some drawbacks.

I think this is still a weakness of TCA, but not sure whether it can be fixed in a cleaner way (ObservedObject has the habit of otherwise refreshing the whole view body).

**Reply by rzulkoski** [reply]
Thanks for the response @fabstu, although I think this situation is a bit different. I'm specifically referring to Apple's [SwiftUI.Section](https://developer.apple.com/documentation/swiftui/section). My hunch is that when any of the generic views changes, such as the `Footer`, the entire `Section` view is recreated.

**Reply by rzulkoski** [reply]
Okay, I've peeled back another layer of the onion. It appears that `Section` is only misbehaving because my footer is a `Group` conditionally containing different views:

```Swift
} footer: {
  Group { // <--- Wrong, don't do this!
    if someCondition {
      ViewA()
    else {
      ViewB()
    }
  }
  .someSharedModifiers()
}
```

If I replace the `Group` with a `VStack`, the issue where the entire `Section` is recomputed goes away.
